exporter/containerimage/export.go

```diff
@@ -71,108 +71,38 @@ func New(opt Opt) (exporter.Exporter, error) {
 func (e *imageExporter) Resolve(ctx context.Context, opt map[string]string) (exporter.ExporterInstance, error) {
 	i := &imageExporterInstance{
 		imageExporter: e,
-		opts: ImageCommitOpts{
+		opts: opts.ImageCommitOpts{
 			RefCfg: cacheconfig.RefConfig{
 				Compression: compression.New(compression.Default),
 			},
 			BuildInfo:   true,
-			Annotations: make(AnnotationsGroup),
+			Annotations: make(opts.AnnotationsGroup),
 		},
 		store: true,
 	}
 
-	opt, err := i.opts.Load(opt)
-	if err != nil {
-		return nil, err
-	}
+	x := opts.NewExtractor(opt)
+	i.opts.Load(x)
 
-	for k, v := range opt {
-		switch k {
-		case keyPush:
-			if v == "" {
-				i.push = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.push = b
-		case keyPushByDigest:
-			if v == "" {
-				i.pushByDigest = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.pushByDigest = b
-		case keyInsecure:
-			if v == "" {
-				i.insecure = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.insecure = b
-		case keyUnpack:
-			if v == "" {
-				i.unpack = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.unpack = b
-		case keyStore:
-			if v == "" {
-				i.store = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.store = b
-		case keyUnsafeInternalStoreAllowIncomplete:
-			if v == "" {
-				i.storeAllowIncomplete = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.storeAllowIncomplete = b
-		case keyDanglingPrefix:
-			i.danglingPrefix = v
-		case keyNameCanonical:
-			if v == "" {
-				i.nameCanonical = true
-				continue
-			}
-			b, err := strconv.ParseBool(v)
-			if err != nil {
-				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
-			}
-			i.nameCanonical = b
-		default:
-			if i.meta == nil {
-				i.meta = make(map[string][]byte)
-			}
-			i.meta[k] = []byte(v)
-		}
+	x.ExtractBoolDefault(opts.OptKey(keyPush), &i.push, true)
```

Comment:
I'm not so sure about this one. The old code is more easier to understand as the logic is quite simple and not very hard to follow. New one adds a new intermediary library and anyone reading this or trying to make a modification needs to understand what methods this custom package supports and how is it intended to be used. `RestBytes()/Error()/OptKeyer` do not seem to follow any common convention either.

And if you look at the line count then it is actually adding more lines.

Reply:
Agreed, this didn't come out quite out quite how I liked it - will close, and can always revisit later if necessary :+1: